src/ethereum/osaka/bal_utils.py

```diff
@@ -0,0 +1,338 @@
+"""
+BAL Utilities for EIP-7928
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Utilities for working with Block Access Lists, including hashing and validation.
+"""
+
+from typing import Union, Optional
+from ethereum_types.bytes import Bytes
+from ethereum_types.numeric import Uint
+
+from ethereum.crypto.hash import Hash32, keccak256
+
+from .ssz_types import (
+    BlockAccessList,
+    AccountChanges,
+    SlotChanges,
+    SlotRead,
+    StorageChange,
+    BalanceChange,
+    NonceChange,
+    CodeChange,
+    MAX_TXS,
+    MAX_SLOTS,
+    MAX_ACCOUNTS,
+    MAX_CODE_SIZE,
+)
+
+
+def compute_bal_hash(bal: BlockAccessList) -> Hash32:
+    """
+    Compute the hash of a Block Access List.
+    
+    The BAL is SSZ-encoded and then hashed with keccak256.
+    
+    Parameters
+    ----------
+    bal :
+        The Block Access List to hash.
+        
+    Returns
+    -------
+    hash :
+        The keccak256 hash of the SSZ-encoded BAL.
+    """
+    bal_bytes = ssz_encode_block_access_list(bal)
+    return keccak256(bal_bytes)
+
+
+def ssz_encode_uint(value: Union[int, Uint], size: int) -> bytes:
+    """Encode an unsigned integer as SSZ (little-endian)."""
+    if isinstance(value, Uint):
+        value = int(value)
+    return value.to_bytes(size, 'little')
+
+
+def ssz_encode_bytes(data: bytes) -> bytes:
+    """Encode fixed-size bytes as SSZ."""
+    return data
+
+
+def ssz_encode_list(items: tuple, encode_item_fn, max_length: int = None) -> bytes:
+    """Encode a list/tuple as SSZ with optional max length."""
+    # For variable-length lists, we need offset encoding
+    # First, encode the list length
+    result = bytearray()
+    
+    # If max_length is specified, this is a variable-length list
+    if max_length is not None:
```

Comment:
Instead of negating here, flip the order of the branches (so the condition becomes `max_length is None` and the else block is moved up).